### PR TITLE
Simple lexer for formatter

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR introduces a simple zero-allocation lexer specific for the needs of the formatter.
It supports a very limited set of tokens for optimized performance.

The main reason for introducing a lexer is that Charlie correctly pointed out that the current implementation for testing if an expression is parenthesized is not sufficient. 
Having a more proper lexer will simplify testing for the right tokens. 

## Test Plan

`cargo test`, new unit tests

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4862** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4862" target="_blank">Graphite</a> 
    * **PR #4895** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4895" target="_blank">Graphite</a> 
      * **PR #4904** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4904" target="_blank">Graphite</a> 
        * **PR #4921** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4921" target="_blank">Graphite</a> 
          * **PR #4922** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4922" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/charliermarsh/ruff/4922?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/trivia.rs [hunk 1]
```diff
@@ -9,174 +10,630 @@ use ruff_text_size::{TextLen, TextRange, TextSize};
...
+    Colon,
+
+    /// '/'
+    Slash,
+
+    /// Any other non trivia token. Always has a length of 1
+    Other,
+
+    /// Returned for each character after [`TokenKind::Other`] has been returned once.
+    Bogus,
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/trivia.rs:167]: Should we name this `Unknown` instead?

File: crates/ruff_python_formatter/src/trivia.rs [hunk 2]
```diff
@@ -9,174 +10,643 @@ use ruff_text_size::{TextLen, TextRange, TextSize};
...
+        .skip_trivia()
+        .next_back()
+}
+
+/// Returns the number of newlines between `offset` and the first non whitespace character in the source code.
+pub(crate) fn lines_before(offset: TextSize, code: &str) -> u32 {
+    let tokens = SimpleTokenizer::up_to(offset, code);
+    let mut newlines = 0u32;
+
+    for token in tokens.rev() {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/trivia.rs:36]: One of the main benefits is that we no longer repeat the same logic over and over again (and e.g. correctly handle continuation tokens)

**Comment by MichaReiser** [comment]
@charliermarsh let me know if you have time to look at this PR. I'll otherwise go ahead and merge it, as there's already a lot of code depending on it

**Comment by charliermarsh** [comment]
@MichaReiser - Go for it, I won't have time for a few hours.

**Pull request merged**